src/test/java/org/jenkinsci/plugins/workflow/graph/FlowNodeTest.java

```diff
@@ -454,6 +457,34 @@ public void evaluate() throws Throwable {
         assertThat(nodes, hasSize(2));
         assertWarning(nodes.get(0), Result.FAILURE, BallColor.RED);
         assertWarning(nodes.get(1), Result.UNSTABLE, BallColor.YELLOW);
+        });
+    }
+
+    @Issue("JENKINS-57805")
+    @Test
+    public void nodeWithNoParentsInBruteForceScanForEnclosingBlock() throws Exception {
+        logging.capture(10);
+        rr.thenWithHardShutdown(j -> {
+            WorkflowJob job = j.jenkins.createProject(WorkflowJob.class, "p");
+            job.setDefinition(new CpsFlowDefinition(
+                    "echo 'to-corrupt'\n" +
+                    "sleep 1\n" + // Trigger a save
+                    "semaphore 'marker'", true));
+            WorkflowRun run = job.scheduleBuild2(0).waitForStart();
+            SemaphoreStep.waitForStart("marker/1", run);
+            CpsFlowExecution cpsExec = (CpsFlowExecution)run.getExecution();
+            // Corrupt a flow node so that we get an error when trying to load it.
+            Files.write(cpsExec.getStorageDir().toPath().resolve("3.xml"), "garbage".getBytes(StandardCharsets.UTF_8));
+        });
+        rr.then(j -> {
+            WorkflowJob job = j.jenkins.getItemByFullName("p", WorkflowJob.class);
+            WorkflowRun run = job.getBuildByNumber(1);
+            CpsFlowExecution cpsExec = (CpsFlowExecution)run.getExecution();
+            // We expect to see a warning in the logs, but we want to make sure that an IndexOutOfBoundsException
+            // is not thrown out of StandardGraphLookupView.bruteForceScanForEnclosingBlockSafety
+            assertThat(cpsExec.getCurrentHeads().get(0).getEnclosingBlocks(), hasSize(0));
+            assertThat(logging.getMessages(), hasItem("failed to load parents of 4"));
```

Comment:
This comes from the warning printed to the logs by `FlowNode.loadParents` [here](https://github.com/jenkinsci/workflow-api-plugin/blob/6da45c1339cf2b9bdd9c7f79d08b6caa827af2d9/src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java#L173).